### PR TITLE
added self organising maps algorithm  in the machine learning section.

### PR DESCRIPTION
**Pull request by Eeman1113**
### Describe your change:

* [x] Add an algorithm?
* [ ] Fix a bug or typo in an existing algorithm?
* [ ] Documentation change?

### Checklist:
* [ ] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md).
* [ ] This pull request is all my own work -- I have not plagiarized.
* [ ] I know that pull requests will not be merged if they fail the automated tests.
* [ ] This PR only changes one algorithm file.  To ease review, please open separate PRs for separate algorithms.
* [ ] All new Python files are placed inside an existing directory.
* [ ] All filenames are in all lowercase characters with no spaces or dashes.
* [ ] All functions and variable names follow Python naming conventions.
* [ ] All function parameters and return values are annotated with Python [type hints](https://docs.python.org/3/library/typing.html).
* [ ] All functions have [doctests](https://docs.python.org/3/library/doctest.html) that pass the automated testing.
* [ ] All new algorithms have a URL in its comments that points to Wikipedia or other similar explanation.
* [ ] If this pull request resolves one or more open issues then the commit message contains `Fixes: #{$ISSUE_NO}`.

### CODE REVIEW

File: machine_learning/Self_Organising_Maps.py [hunk 1]
```diff
@@ -0,0 +1,62 @@
+import math
+
+
+class SOM:
+    # Function here computes the winning vector
+    # by Euclidean distance
+    def winner(self, weights, sample):
```
[the remark this answers is not in the document]
Comment (Eeman1113) [reply machine_learning/Self_Organising_Maps.py:7]: The output for winner is in int format 
All the inputs for winner is in list format

weight - int type 
sample - int type

Comment (cclauss) [inline machine_learning/Self_Organising_Maps.py:7]: These are type hints...
```suggestion
    def winner(self, weights: int, sample: int) --> list[int]:
```

File: machine_learning/self_organizing_map.py [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+        >>> SelfOrganizingMap().update([[1, 2, 3], [4, 5, 6]], [1, 2, 3], 1, 0.1)
+        [[1, 2, 3], [3.7, 4.7, 6]]
+        """
+        for i in range(len(weights)):
+            weights[j][i] += alpha * (sample[i] - weights[j][i])
+        return weights
+
+
+# Driver code
+def main() -> None:
```
[the remark this answers is not in the document]
Comment (cclauss) [reply machine_learning/self_organizing_map.py:38]: @dhruvmanila Does the `main()` function _really_ need doctests?

**Review by cclauss — APPROVED** [review]
As discussed in `CONTRIBUTING.md`, I added type hints, doctests, and better names for functions and variables while trying to keep the algorithm unchanged.

**Pull request merged**